File: packages/codeality-py/src/codeality_py/gate/gate_stages.py
"""The stages the gate runs, in order."""

from codeality_py.baseline.baseline_file import BASELINE_FILENAME
from codeality_py.config.baseline_config import BaselineConfig
from codeality_py.gate.stage import Stage
from codeality_py.gate.stage_kind import StageKind


def gate_stages(config: BaselineConfig) -> tuple[Stage, ...]:
    """Return the ordered stages for this project.

    Coverage is collected rather than merely configured. pip-audit audits the
    installed environment, which the CI workflow builds with "uv sync
    --locked", so what is audited is exactly what the lockfile pins.

    pytest reports its slowest tests every run, so a suite that goes over its
    budget hands the reader the measurement in the same output, and a suite
    creeping towards it is visible before it gets there.
    """
    roots = [*config.source_roots, *config.test_roots]
    package = config.import_package or config.source_roots[0]
    # A recorded baseline is the migration plan; the gate must honor it, or
    # recording debt could never turn the gate green.
    structural = (
        ("codeality-py", "baseline", "check")
        if (config.project_root / BASELINE_FILENAME).is_file()
        else ("codeality-py", "check")
    )
    return (
        Stage("ruff", StageKind.REQUIRED, ("ruff", "check", *roots)),
        Stage("ruff-format", StageKind.REQUIRED, ("ruff", "format", "--check", *roots)),
        Stage("mypy", StageKind.REQUIRED, ("mypy", *config.source_roots)),
        Stage("codeality-py", StageKind.REQUIRED, structural),
        Stage("deptry", StageKind.REQUIRED, ("deptry", ".")),
        Stage(
            "pip-audit",
            StageKind.REQUIRED,
            (
                "pip-audit",
                *(part for vuln in config.audit_ignore_vulns for part in ("--ignore-vuln", vuln)),
            ),
        ),
        Stage(
            "pytest",
            StageKind.REQUIRED,
            (
                "pytest",
                f"--cov={package}",
                "--cov-report=term-missing",
                f"--cov-fail-under={config.coverage_threshold}",
                "--durations=10",
            ),
            budget_seconds=float(config.test_budget_seconds),
        ),
        Stage("pyrefly", StageKind.SHADOW, ("pyrefly", "check")),
    )

File: packages/db-quality/src/gate/gateStages.ts
import { existsSync } from 'node:fs'
import { join } from 'node:path'

import { runAudit } from '@/audit/runAudit.js'
import { BASELINE_FILENAME } from '@/baseline/BASELINE_FILENAME.js'
import { classifyFindings } from '@/baseline/classifyFindings.js'
import { readBaseline } from '@/baseline/readBaseline.js'
import type { CheckContext } from '@/check/CheckContext.js'
import { runCheck } from '@/check/runCheck.js'
import { perfStage } from '@/gate/perfStage.js'
import type { Stage } from '@/gate/Stage.js'

// A recorded baseline is the migration plan; the gate must honour it, or
// recording debt could never turn the gate green. Audit runs when the project
// is linked unless the configuration says otherwise: a linked project whose
// CI has no token must switch it off explicitly rather than pass by accident.
export const gateStages = (context: CheckContext): Stage[] => {
  const hasBaseline = existsSync(join(context.root, BASELINE_FILENAME))
  const linked = existsSync(join(context.root, 'supabase/.temp/project-ref'))
  return [
    hasBaseline
      ? {
          name: 'baseline-check',
          run: () =>
            classifyFindings(runCheck(context), readBaseline(context.root)).new,
        }
      : { name: 'check', run: () => runCheck(context) },
    {
      name: 'audit',
      run: () =>
        context.config.audit.inGate && linked
          ? runAudit({ ...context, target: { linked: true } })
          : 'not-applicable',
    },
    perfStage(context),
  ]
}
